core/storage/gcp_native_chunk_manager.go

```diff
@@ -0,0 +1,472 @@
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (gcm *GCPNativeChunkManager) RemoveWithPrefix(ctx context.Context, bucketName string, prefix string) error {
+	objects := gcm.client.Bucket(bucketName).Objects(ctx, &storage.Query{Prefix: prefix})
+
+	// Group objects by their depth (number of / in the key)
+	groupedByLevel := make(map[int][]string)
+	var maxLevel int
+	for {
+		objAttrs, err := objects.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		level := strings.Count(objAttrs.Name, "/")
+		groupedByLevel[level] = append(groupedByLevel[level], objAttrs.Name)
+		if level > maxLevel {
+			maxLevel = level
+		}
+	}
+
+	for level := maxLevel; level >= 0; level-- {
+		// Get the objects at this level
+		keysAtLevel := groupedByLevel[level]
+		if len(keysAtLevel) == 0 {
+			continue
+		}
+		// Dynamically adjust maxGoroutines based on the number of objects at this level
+		maxGoroutines := 10
+		if len(keysAtLevel) < maxGoroutines {
+			maxGoroutines = len(keysAtLevel)
+		}
+		i := 0
+		for i < len(keysAtLevel) {
+			runningGroup, groupCtx := errgroup.WithContext(context.Background())
+			for j := 0; j < maxGoroutines && i < len(keysAtLevel); j++ {
+				key := keysAtLevel[i]
+				runningGroup.Go(func() error {
+					err := gcm.Remove(groupCtx, bucketName, key)
+					if err != nil {
+						log.Warn("failed to remove object", zap.String("bucket", bucketName), zap.String("path", key), zap.Error(err))
+						return err
+					}
+					return nil
+				})
+				i++
+			}
+			if err := runningGroup.Wait(); err != nil {
+				return err
+			}
+		}
+	}
+	err := gcm.Remove(ctx, bucketName, strings.TrimSuffix(prefix, "/"))
+	if err != nil {
+		log.Warn("failed to remove object", zap.String("bucket", bucketName), zap.String("path", prefix), zap.Error(err))
+		return err
+	}
+	return nil
+}
+
+// Copy files from fromPath into toPath recursively
+func (gcm *GCPNativeChunkManager) Copy(ctx context.Context, fromBucketName string, toBucketName string, fromPath string, toPath string) error {
+	bkt := gcm.client.Bucket(fromBucketName)
+	it := bkt.Objects(ctx, &storage.Query{Prefix: fromPath})
+	for {
+		objAttrs, err := it.Next()
+		if err == iterator.Done {
+			break // No more objects to copy
+		}
+		if err != nil {
+			log.Error("Error listing objects", zap.Error(err))
+			return err
+		}
+
+		srcObjectKey := objAttrs.Name
+		dstObjectKey := strings.Replace(srcObjectKey, fromPath, toPath, 1)
+
+		srcReader, err := bkt.Object(srcObjectKey).NewReader(ctx)
+		if err != nil {
+			log.Error("Error creating reader for object", zap.String("srcObjectKey", srcObjectKey), zap.Error(err))
+			return err
+		}
+
+		defer func() {
+			if err := srcReader.Close(); err != nil {
+				log.Error("Error closing source reader", zap.String("srcObjectKey", srcObjectKey), zap.Error(err))
+			}
+		}()
+
+		dstWriter := gcm.client.Bucket(toBucketName).Object(dstObjectKey).NewWriter(ctx)
```

Comment:
Use defer func() for `dstWriter` as well for constancy.

Reply:
sure updated